File: 3569-count-of-substrings-containing-every-vowel-and-k-consonants-ii/3569-count-of-substrings-containing-every-vowel-and-k-consonants-ii.go
package main

import (
	"fmt"
)

func countOfSubstrings(word string, k int) int64 {
	n := len(word)
	res := int64(0)
	vowels := map[byte]bool{'a': true, 'e': true, 'i': true, 'o': true, 'u': true}
	suf := make([]int, n+1)

	for i := n - 1; i >= 0; i-- {
		if i+1 < n && vowels[word[i+1]] {
			suf[i] = suf[i+1] + 1
		} else {
			suf[i] = 0
			if vowels[word[i]] {
				suf[i] = 1
			}
		}
	}

	freq := make(map[byte]int)
	for i, j, cnt := 0, 0, 0; i < n; i++ {
		for j < n && (len(freq) != 5 || cnt < k) {
			if vowels[word[j]] {
				freq[word[j]]++
			} else {
				cnt++
			}
			j++
		}
		if len(freq) == 5 && cnt == k {
			res += int64(max(1, suf[j-1]))
		}
		if vowels[word[i]] {
			freq[word[i]]--
			if freq[word[i]] == 0 {
				delete(freq, word[i])
			}
		} else {
			cnt--
		}
	}

	return res
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: 3569-count-of-substrings-containing-every-vowel-and-k-consonants-ii/3569-count-of-substrings-containing-every-vowel-and-k-consonants-ii.cpp
class Solution {
public:
    long long countOfSubstrings(string word, int k) {
        long long n = word.length(), res = 0;
        unordered_set<char> vowels{'a','e','i','o','u'};
        vector<int> suf(n + 1);
        for(int i = n - 1; i >= 0; i--) {
            if(i + 1 < n and vowels.count(word[i+1])) suf[i] = suf[i+1] + 1;
            else suf[i] = vowels.count(word[i]);
        }
        unordered_map<char,int> freq;
        for(int i = 0, j = 0, cnt = 0; i < n; i++) {
            while(j < n and (freq.size() != 5 or cnt < k)) {
                if(vowels.count(word[j])) freq[word[j]]++;
                else cnt++;
                j++;
            }
            if(freq.size() == 5 and cnt == k) {
                res += max(1, suf[j-1]);
            }
            if(vowels.count(word[i])) {
                if(--freq[word[i]] == 0) freq.erase(word[i]);
            } else cnt--;
        }
        return res;
    }
};